### PR TITLE
DRUP-420 : Add docstore API integration to Apigee Client for PHP library

### PR DESCRIPTION
**Pull request by giteshk**
Allows the ability to interact with the Docstore apis. 
Docstore is only available on-cloud today.

### CODE REVIEW

File: src/Api/Docstore/Serializer/DocstoreSerializer.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace Apigee\Edge\Api\Docstore\Serializer;
+
+use Apigee\Edge\Api\Docstore\Denormalizer\DocstoreDenormalizer;
+use Apigee\Edge\Serializer\EntitySerializer;
+
+class DocstoreSerializer extends EntitySerializer
+{
+    /**
+     * CollectionSerializer constructor.
```
Comment (mxr576) [inline src/Api/Docstore/Serializer/DocstoreSerializer.php]: Outdated doc block comment.

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+use Apigee\Edge\Api\Docstore\Entity\Folder;
+use Apigee\Edge\Api\Docstore\Serializer\DocstoreSerializer;
+use Apigee\Edge\Controller\EntityController;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * Class DocstoreController allows for CRUD operations for Docstore entities (Folder and Doc).
+ */
+class DocstoreController extends EntityController
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: Please create an interface for this controller and move all public method definitions to that.

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    /**
+     * @inheritdoc
+     */
+    public function create(DocstoreObject $entity): void
+    {
+        if (null == $entity->getFolder()) {
+            $entity->setFolder($this->getHomeFolderId());
+        }
+        $response = $this->getClient()->post(
+            ($entity instanceof Doc ? 'specs/new' : 'folders'),
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: Let's separate this logic to dedicated createFolder() and createSpec() methods. Controller methods should serve only one purpose.

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+            $this->getEntitySerializer()->serialize($entity, 'json'),
+            $this->getHeaders());
+        $this->getEntitySerializer()->setPropertiesFromResponse($response, $entity);
+        $entity->setEtag($response->getHeader('ETag')[0]);
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: The entity serializer should be the one that parses _all_ information from an API response and creates/updates an entity. So etag should be set by the serializer also.

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+     */
+    public function delete(string $entityId): DocstoreObject
+    {
+        $response = $this->getClient()->delete($this->getEntityEndpointUri($entityId), null, $this->getHeaders());
+
+        return $this->parseDocstoreResponse($response);
+    }
+
+    /**
+     * Upload a openAPI file to the backend.
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: s/openApi/OpenAPI/g

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+     *
+     * This is what the API supports today.
+     * Uploading a JSON file and setting the content-type to application/x-yaml
+     * This will change in the future
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: Missing dot.

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+        $this->getClient()->put(
+            $entity->getContent(),
+            $content,
+            $this->getHeaders() + ['Content-Type' => 'application/x-yaml']);
+    }
+
+    /**
+     * Get the contents of the spec.
+     *
+     * Always returns application/json
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: Missing dot.

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+            $homeFolderId = $homeFolder->id();
+        }
+
+        return $homeFolderId;
+    }
+
+    /**
+     * Parses the DocStore response and generates either a Folder/Doc object.
+     */
+    private function parseDocstoreResponse(ResponseInterface $response): DocstoreObject
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: This is the task of an entity serializer and not a controller. Precisely the denormalizers inside an entity serializer should decide what object they return based on an API response. Please remove this logic here.

File: src/Api/Docstore/Entity/Collection.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Apigee\Edge\Api\Docstore\Entity;
+
+/**
+ * Collection object in Docstore.
+ */
+class Collection extends \Apigee\Edge\Api\Docstore\Entity\DocstoreObject
```
Comment (mxr576) [inline src/Api/Docstore/Entity/Collection.php]: Do not use FQCN.

Comment (mxr576) [reply src/Api/Docstore/Entity/Collection.php]: Are you sure that this is an actual entity and not just a structure?

File: src/Api/Docstore/Entity/Collection.php [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+{
+    protected $kind = 'Collection';
```
Comment (mxr576) [inline src/Api/Docstore/Entity/Collection.php:29]: As I can see there 1:1 relationship between the "kind" and the class so there is no need to store the kind as property in object - especially it should not be defined in all sub classes of the DocsStore class - rather the denormalizer could use the "kind" information from the API response to create an object with a proper type and the normalizer could add this information (property) to the API request payload based on the type of the normalized class before it gets sent to Apigee Edge. (See Rate plans as an example.)

File: src/Api/Docstore/Entity/Doc.php [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Apigee\Edge\Api\Docstore\Entity;
+
+/**
+ * Spec object from the specstore.
+ */
+class Doc extends DocstoreObject
+{
+    /**
+     * url to fetch the spec file.
```
Comment (mxr576) [inline src/Api/Docstore/Entity/Doc.php]: Sentence starts with a capital letter.

File: src/Api/Docstore/Entity/DocstoreObject.php [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+use Apigee\Edge\Entity\Property\NamePropertyAwareTrait;
+
+/**
+ * Abstract object to represent the Docstore entities.
+ */
+abstract class DocstoreObject extends Entity
+{
+    use NamePropertyAwareTrait;
+    use DescriptionPropertyAwareTrait;
+    use \Apigee\Edge\Entity\CommonEntityPropertiesAwareTrait;
```
Comment (mxr576) [inline src/Api/Docstore/Entity/DocstoreObject.php]: Do not use FQCN.

File: src/Api/Docstore/Entity/DocstoreObject.php [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+
+    protected $permissions;
```
Comment (mxr576) [inline src/Api/Docstore/Entity/DocstoreObject.php]: Missing typehint from properties, some of them may deserve a description as well.

File: src/Api/Docstore/Entity/DocstoreObject.php [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    {
+        $this->permissions = $permissions;
+    }
+
+    /**
+     * Is the specstore object trashed.
+     *
+     * @return mixed
+     */
+    public function getIsTrashed()
```
Comment (mxr576) [inline src/Api/Docstore/Entity/DocstoreObject.php]: This is a boolean property so its getter should `isTranshed()` and it setter should be `setTrashed()`. 
See Address' name converter as an example for how you can fix this:
https://github.com/apigee/apigee-client-php/blob/2.x/src/Api/Monetization/NameConverter/AddressNameConverter.php#L31
https://github.com/apigee/apigee-client-php/blob/2.x/src/Api/Monetization/Denormalizer/AddressDenormalizer.php#L41
https://github.com/apigee/apigee-client-php/blob/2.x/src/Api/Monetization/Normalizer/AddressNormalizer.php#L41
https://github.com/apigee/apigee-client-php/blob/2.x/src/Api/Monetization/Structure/Address.php#L29

File: src/Api/Docstore/Entity/DocstoreObject.php [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+namespace Apigee\Edge\Api\Docstore\Entity;
+
+use Apigee\Edge\Entity\Entity;
+use Apigee\Edge\Entity\Property\DescriptionPropertyAwareTrait;
+use Apigee\Edge\Entity\Property\NamePropertyAwareTrait;
+
+/**
+ * Abstract object to represent the Docstore entities.
+ */
+abstract class DocstoreObject extends Entity
```
Comment (mxr576) [inline src/Api/Docstore/Entity/DocstoreObject.php]: Please define an interface for this class, move all its public definition to that class and please define interfaces for all its subclasses as well.

File: tests/Api/Docstore/Controller/DocstoreControllerTest.php [hunk 16]
```diff
@@ -0,0 +1,664 @@
...
+use Monolog\Formatter\LineFormatter;
+use Monolog\Handler\StreamHandler;
+use Monolog\Logger;
+use Monolog\Processor\PsrLogMessageProcessor;
+
+/**
+ * Class DocstoreControllerTest.
+ *
+ * @group controller
+ * @group management
```
Comment (mxr576) [inline tests/Api/Docstore/Controller/DocstoreControllerTest.php]: This test does not belong to a management API test, this is what this group stands for. I'd rather introduce a new `docstore`.

File: tests/Api/Docstore/Controller/DocstoreControllerTest.php [hunk 17]
```diff
@@ -0,0 +1,664 @@
...
+    /**
+     * Test to create a Sample doc and verify it was created.
+     */
+    public function testSampleSpecExists(): void
+    {
+        static::markOnlineTestSkipped(__FUNCTION__);
+        $spec = new Doc(['name' => static::getSampleDocName()]);
+        static::entityController()->create($spec);
+
+        $homeFolder = static::entityController()->load('/homeFolder');
```
Comment (mxr576) [inline tests/Api/Docstore/Controller/DocstoreControllerTest.php:96]: Why `/` is necessary? It should not be. `load()` method on an entity controller loads an entity by id.

File: tests/Api/Docstore/Controller/DocstoreControllerTest.php [hunk 18]
```diff
@@ -0,0 +1,664 @@
...
+    use MarkOnlineTestSkippedAwareTrait;
+    use DefaultAPIClientAwareTrait;
+
+    /**
+     * Delete all the generated test data.
+     */
+    public static function tearDownAfterClass(): void
+    {
+        parent::tearDownAfterClass();
+        static::removeTestData();
```
Comment (mxr576) [inline tests/Api/Docstore/Controller/DocstoreControllerTest.php:61]: Feel free to inline the method content's here. It seems that you only use this in one place.

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+    }
+
+    /**
+     * @inheritdoc
+     *
+     * @psalm-suppress PossiblyNullArgument $entity->id() is always not null here.
+     */
+    public function update(DocstoreObject $entity): void
+    {
+        $uri = $this->getEntityEndpointUri($entity->id());
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: Unnecessary new variable `$uri`, you only use it in one place.

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+        $update_arr = [];
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: Use camelCase variable names inside the PHP API Client.

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+        $update_arr['folder'] = $entity->getFolder();
+        $update_arr['name'] = $entity->getName();
+        if (null !== $entity->getDescription()) {
+            $update_arr['description'] = $entity->getDescription();
+        }
+        $update_arr['isTrashed'] = $entity->getIsTrashed();
+        // Update an existing entity.
+        $response = $this->getClient()->patch(
+            $uri,
+            $this->getEntitySerializer()->serialize($update_arr, 'json'),
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: Can't you pass the entity directly to the serializer that should generate the required request payload from it? If not, why?

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 22]
```diff
@@ -0,0 +1,271 @@
...
+     * @param Doc $entity
+     *
+     * @throws \Http\Client\Exception
+     *
+     * @return string
+     */
+    public function getSpecContents(Doc $entity): string
+    {
+        $response = $this->getClient()->get($entity->getContent(),
+            $this->getHeaders() + ['Accept' => 'application/json, text/plain, */*']);
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php]: Isn't only "application/json" what this API can return? If not "text/plan" could be acceptable if the API returns a JSON file with incorrect content type, but I do not understand why "*/*" is necessary. Basically it means we accept anything, although our serializer can only read JSON.

**Comment by mxr576** [comment]
No offense, but I do not get how this code passed on the code style to be honest, probably the code style checker requires some adjustments ;)
* variable names should be camelCase
* At least variable type hint should be added to the class properties, one new line between each property.
* All entity and controller classes should have an interface defined that describes its public behaviors - this is something that we can not validate with static code analyses at this moment.

File: src/Api/Docstore/Controller/DocstoreController.php [hunk 23]
```diff
@@ -0,0 +1,271 @@
...
+
+        return $collectionSerializer->deserialize(
+            (string) $response->getBody(),
+            Collection::class,
+            'json'
+        );
+    }
+
+    /**
+     * Docstore entity id start with a "/".
```
Comment (mxr576) [inline src/Api/Docstore/Controller/DocstoreController.php:266]: Entity ids should not start with `/`. Just add that to the path builder...
https://github.com/apigee/apigee-client-php/blob/8017c3993ab78c239744b4b1fdd25deebf61678b/src/Api/Management/Controller/DeveloperAppCredentialController.php#L67

File: src/Api/Docstore/Entity/Collection.php (hunk 9 above)
Comment (mxr576) [reply src/Api/Docstore/Entity/Collection.php]: What could be in a collection? Folders? Docs? Can they be mixed or only from the same type? If the answer is one type per collection then I think you should introduce new FolderCollection and DocCollection classes.

**Review by mxr576 — CHANGES_REQUESTED** [review]

File: src/Api/Docstore/Controller/DocstoreController.php (hunk 21 above)
Comment (giteshk) [reply src/Api/Docstore/Controller/DocstoreController.php]: The patch method only takes certain values. I tried passing the entire entity and it fails.

File: src/Api/Docstore/Controller/DocstoreController.php (hunk 22 above)
Comment (giteshk) [reply src/Api/Docstore/Controller/DocstoreController.php]: This is the actual openAPI specification that is uploaded in edge.

The API documentation had that in the instructions. But I have tested this with application/json and my tests ran. So I am setting this to only be application/json

File: src/Api/Docstore/Controller/DocstoreController.php (hunk 23 above)
Comment (giteshk) [reply src/Api/Docstore/Controller/DocstoreController.php:266]: Entity ID returned by the backend API is in this format "/c3Rvc-ZG9j-93459"

The logic to generate the entity URL is abstracted in this one method.

File: src/Api/Docstore/Entity/Collection.php (hunk 9 above)
Comment (giteshk) [reply src/Api/Docstore/Entity/Collection.php]: Collection is the object that is returned when you pull in contents from any folder

File: tests/Api/Docstore/Controller/DocstoreControllerTest.php (hunk 17 above)
Comment (giteshk) [reply tests/Api/Docstore/Controller/DocstoreControllerTest.php:96]: entity id's for all docstore entities begin with a "/"

File: src/Api/Docstore/Entity/Collection.php (hunk 10 above)
Comment (giteshk) [reply src/Api/Docstore/Entity/Collection.php:29]: Need this when running the create operation on the docstore entities.

Either I do this or have to set the kind in the constructor when I initialize it.

File: src/Api/Docstore/Entity/DocstoreObject.php (hunk 14 above)
Comment (giteshk) [reply src/Api/Docstore/Entity/DocstoreObject.php]: I implemented this. 

BTW the trashed flag will always return false. There is no concept of Trash in docstore as it stands today.  I think they might implement this in the future.

**Comment by giteshk** [comment]
I think there was a merge issue. I will close this and create a new pull request.

**Pull request closed without merging**